### ISSUE TITLE
ANR issue executing service com.*/com.onesignal.SyncService

### ISSUE THREAD

**Issue by BobyIlea** [OneSignal/OneSignal-Android-SDK#283]
We are getting this ANR report very often on Google Play console. I have searched for answers regarding this issue and apparently it should have been fixed in v3.3.1, but we are still getting this ANR in latest builds which are currently using v3.5.8

From build.gradle:
compile 'com.onesignal:OneSignal:3.5.8'
compile 'com.android.support:appcompat-v7:24.+'
compile 'com.google.android.gms:play-services-gcm:11.0.2'
compile 'com.google.android.gms:play-services-location:11.0.2'

I've also attached the logs. Please let me know if there is anything else that you need from our side.

[log_ft_63.txt](https://github.com/OneSignal/OneSignal-Android-SDK/files/1234051/log_ft_63.txt)

**Comment by BobyIlea** [comment]
Hello, did anyone have time to look into this? Thank you.

**Comment by ron-k** [comment]
Seeing here same ANR.
Using com.onesignal:OneSignal:3.5.8

I have the fix, please see the references PR.

**Comment by BobyIlea** [comment]
@ron-k Thanks, do you know if your fix has been merged in a later release? We're not using the sources to build the OneSignal library.

**Comment by jkasten2** [comment]
@BobyIlea This specific ANR should be fixed in the now released 3.6.2 SDK.

**Comment by BobyIlea** [comment]
@jkasten2 Thank you for your time!

**Comment by DimaKoz** [comment]
@jkasten2  I'm sorry but I've got ANR when executing service com.my.package/com.onesignal.SyncService (unfortunately, in 3.6.2):
"main" prio=5 tid=1 WAIT
  | group="main" sCount=1 dsCount=0 obj=0x418ecea0 self=0x4180aff0
  | sysTid=10864 nice=-11 sched=0/0 cgrp=apps handle=1074401620
  | state=S schedstat=( 0 0 0 ) utm=101 stm=9 core=0
 
  at java.lang.Object.wait (Native Method)
- waiting on <0x418ecf70> (a java.lang.VMThread) held by tid=1 (main)
  at java.lang.Thread.parkFor (Thread.java:1205)
  at sun.misc.Unsafe.park (Unsafe.java:325)
  at java.util.concurrent.locks.LockSupport.park (LockSupport.java:157)
  at java.util.concurrent.locks.AbstractQueuedSynchronizer.parkAndCheckInterrupt (AbstractQueuedSynchronizer.java:813)
  at java.util.concurrent.locks.AbstractQueuedSynchronizer.doAcquireSharedInterruptibly (AbstractQueuedSynchronizer.java:973)
  at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquireSharedInterruptibly (AbstractQueuedSynchronizer.java:1281)
  at java.util.concurrent.CountDownLatch.await (CountDownLatch.java:202)
  at android.app.SharedPreferencesImpl$EditorImpl$1.run (SharedPreferencesImpl.java:364)
  at android.app.QueuedWork.waitToFinish (QueuedWork.java:88)
  at android.app.ActivityThread.handleServiceArgs (ActivityThread.java:2995)
  at android.app.ActivityThread.access$2200 (ActivityThread.java:174)
  at android.app.ActivityThread$H.handleMessage (ActivityThread.java:1404)
  at android.os.Handler.dispatchMessage (Handler.java:102)
  at android.os.Looper.loop (Looper.java:146)
  at android.app.ActivityThread.main (ActivityThread.java:5593)
  at java.lang.reflect.Method.invokeNative (Native Method)
  at java.lang.reflect.Method.invoke (Method.java:515)
  at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run (ZygoteInit.java:1283)
  at com.android.internal.os.ZygoteInit.main (ZygoteInit.java:1099)
  at dalvik.system.NativeStart.main (Native Method)

**Comment by BobyIlea** [comment]
@DimaKoz This one seems to have a different cause than the one I've posted logs, do you have steps to reproduce?

**Comment by DimaKoz** [comment]
@BobyIlea I can't reproduce it( I got them via GooglePlay, 685 ANR and 206 users ) but I guess [this issue](https://github.com/OneSignal/OneSignal-Android-SDK/issues/299) is related with that ANR

**Comment by jkasten2** [comment]
@DimaKoz Could be related to `SharedPreferences` as I see this in the log. Does GooglePlay's log give you a full log of all threads?

**Comment by DimaKoz** [comment]
@jkasten2 Sure. 
I noticed that issue has been from OneSignal:3.4.3(as least) to the latest release
[log.txt](https://github.com/OneSignal/OneSignal-Android-SDK/files/1307514/log3.txt)
[log1.txt](https://github.com/OneSignal/OneSignal-Android-SDK/files/1307536/log3.txt)
![screen shot 2017-09-15 at 22 07 16](https://user-images.githubusercontent.com/3092276/30499275-88de4e46-9a62-11e7-9ada-3c7c2565f40f.png)

**Comment by jkasten2** [comment]
@DimaKoz Thanks for the logs. After reviewing them it is an issue with `SharedPreferences` and isn't related to `SyncService`. We made some changes in 3.6.0 that should minimize this. In-addition in the next patch release 3.6.3, we have a for more improvements around writes to `SharedPreferences`.

**Comment by jkasten2** [comment]
@DimaKoz 3.6.3 has been released with the `SharedPreferences` changes. If you continue to see this after this update please open an new issue as OP is about an ANR on location instead.

Thanks.
